docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
 
-For further details about the `.env` migration, read through the
-[`_ss_environment.php` changed to `.env` section](/changelogs/4.0.0#env)
+If you've installed the upgrader, you can use the `recompose` command to help you upgrade your dependencies. This command will try to:
+* upgrade your PHP constraint
+* upgrade core SilverStripe packages to their version 4 equivalent
+* switch to recipes where possible
+* find SilverStripe 4 compatible versions of third party packages.
 
-If you have installed the upgrader tool, you can use the `environment` command to generate a valid `.env` file from your
-existing `_ss_environment.php` file.
+Take for example the following SilverStripe 3 `composer.json` file.
+```json
+{
+    "name": "acme/cms-website",
+    "description": "The ACME website project.",
+    "license": "BSD-3",
+    "require": {
+        "php": ">=5.3.3",
+        "silverstripe/cms": "3.6.5@stable",
+        "silverstripe/framework": "3.6.5@stable",
+        "silverstripe/reports": "3.6.5@stable",
+        "silverstripe/siteconfig": "3.6.5@stable",
+        "dnadesign/silverstripe-elemental": "^1.8.0"
+    }
+}
+```
 
+You can upgrade the `composer.json` file with this command:
+```bash
+/var/www/SS_acme > upgrade-code recompose --recipe-core-constraint=1.1
 ```
-cd ~/my-project-root
-upgrade-code environment --write
+
+The `--recipe-core-constraint` flag can be use to target a specific version of `silverstripe/recipe-core`. If this flag is omitted, the project will be upgraded to the latest stable version. You can use the `--strict` option if you want to use more conservative version constraints. If you want to save your changes, use the `--write` flag,
+
+Your upgraded `composer.json` file will look like this.
+```json
+{
+    "name": "acme/cms-website",
+    "description": "The ACME website project.",
+    "license": "BSD-3",
+    "require": {
+        "dnadesign/silverstripe-elemental": "^2.1",
+        "php": ">=5.6",
+        "silverstripe/recipe-cms": "^1.1"
+    }
+}
 ```
 
-Read the [upgrader `environment` command documentation](https://github.com/silverstripe/silverstripe-upgrader/blob/master/docs/en/environment.md)
-for more details.
+If the `recompose` command can't find a SilverStripe 4 compatible version for one of your package, it will keep this dependency in your `composer.json` file with its existing constraint.
+
+### Manually upgrading your dependencies
+The instruction in this section assumed you'll be editing your `composer.json` file in a text editor.
+
+#### Switching to _recipes_
+Where possible, we recommend you use recipes.
+
+If your SilverStripe 3 project requires the `silverstripe/cms` package, replace that dependency with `silverstripe/recipe-cms`. Set the version constraint for `silverstripe/recipe-cms` to:
+* `~1.0.0` to upgrade to SilverStripe 4.0
+* `~1.1.0` to upgrade to SilverStripe 4.1
+* `~1.2.0` to upgrade to SilverStripe 4.2
```

Comment:
This is completely unintuitive. "To upgrade to SilverStripe 4.0 install recipe 1.0". Find me **one* user who hasn't helped build or maintain the recipe system thinks that this makes sense.

👎 👎 👎 👎 👎 👎 👎 👎 👎

Reply:
@dhensby In this context, if we tell people to use `^1.0.0` to upgrade to 4.0, they're going to be upgraded to 4.2 when they run their `composer update`. 

On the broader point of whatever we should encourage `^` or `~`, it's not clear to me we recommend one or the other. My understanding is that government organisation and bigger business generally prefer using `~` to avoid being accidentally upgraded to a different minor release. In my agency days, I generally preferred to use `^` because I was usually dealing with smaller clients.

The upgrader right now will default to using `^`, unless you set the `--strict` flag.

@sminnee From the discussion on the other retagging issue, looks like we're going to retag. I'll update the instruction to use 4 version when require recipe. However this will block this issue until the retagging is done.